src/malli/roundtrip.clj

```diff
@@ -55,53 +55,139 @@
+(defmethod parsed-overlap? [:fn :orn] [a b]
+  (parsed-overlap? b a))
 (defmethod parsed-overlap? [:map :map] [a b] true)
 (defmethod parsed-overlap? [:vector :set] [a b] false)
 (defmethod parsed-overlap? [:set :set] [a b] true)
 (defmethod parsed-overlap? [:any :any] [a b] true)
 (defmethod parsed-overlap? [:nil :nil] [a b] true)
-(defmethod parsed-overlap? [:number :int] [_ _] true)
-(defmethod parsed-overlap? [:int :number] [_ _] true)
-(defmethod parsed-overlap? [:number :number] [_ _] true)
+;; Predicate overlaps (using symbols as that's how malli represents them)
+(defmethod parsed-overlap? ['number? :int] [_ _] true)
+(defmethod parsed-overlap? [:int 'number?] [_ _] true)
+(defmethod parsed-overlap? ['number? 'number?] [_ _] true)
+(defmethod parsed-overlap? ['int? :int] [_ _] true)
+(defmethod parsed-overlap? [:int 'int?] [_ _] true)
+(defmethod parsed-overlap? ['int? 'int?] [_ _] true)
+(defmethod parsed-overlap? ['int? 'number?] [_ _] true)
+(defmethod parsed-overlap? ['number? 'int?] [_ _] true)
+;; :fn with predicates - for soundness, assume they might overlap
+(defmethod parsed-overlap? [:fn 'int?] [_ _] true)
+(defmethod parsed-overlap? ['int? :fn] [_ _] true)
+(defmethod parsed-overlap? [:fn 'number?] [_ _] true)
+(defmethod parsed-overlap? ['number? :fn] [_ _] true)
+
+(defmulti roundtrippable?
+  (fn [schema] (m/type schema)))
 
 (defn check-child-roundtrip
   [schema path-key]
-  (let [child (second schema)
-        problems (map #(update % :path #(cons path-key %))
-                      (roundtrippable? child))]
+  (let [children (m/children schema)
+        child (first children)
+        problems (vec (map (fn [p] (update p :path (fn [path] (vec (cons path-key path)))))
+                           (roundtrippable? child)))]
     (not-empty problems)))
 
 (defn check-multi-children-roundtrip
   [children]
-  (not-empty
-    (mapcat-indexed (fn [i child]
-                      (map #(update % :path #(cons i %))
-                           (roundtrippable? child)))
-                    children)))
-
-(defmulti roundtrippable?
-  (fn [schema] (m/type schema)))
+  (let [problems (vec
+                  (mapcat (fn [[i child]]
+                            (map (fn [p] (update p :path (fn [path] (vec (cons i path)))))
+                                 (roundtrippable? child)))
+                          (map-indexed vector children)))]
+    (not-empty problems)))
 
 (defn explain
   [path msg & {:as opts}]
   (assoc opts :path path :problem msg))
 
-(doseq [t [:int :string :keyword :boolean :double :uuid :nil :any :enum
-           :number]]
+(doseq [t [:int :string :keyword :boolean :double :uuid :nil :any :enum]]
+  (defmethod roundtrippable? t [_] nil))
+
+;; Predicates are also roundtrippable
+(doseq [t ['int? 'string? 'keyword? 'boolean? 'number? 'uuid? 'nil? 'any? 'double? 'float?
+           'pos-int? 'neg-int? 'nat-int? 'pos? 'neg? 'zero? 'some? 'fn? 'ifn?]]
   (defmethod roundtrippable? t [_] nil))
```

Comment:
Instead of all these explicit cases, can we leverage `-parse-info`? e.g., have a default dispatch that returns nil if `(-> s m/-parse-info :simple-parser)`. This should handle a great many cases, including those from third party libraries that just implement `-parse-info`.